apps/api_web/lib/api_web/controllers/route_pattern_controller.ex

```diff
@@ -93,24 +98,90 @@ defmodule ApiWeb.RoutePatternController do
 
   @spec format_filters(%{optional(String.t()) => String.t()}) :: RoutePattern.filters()
   defp format_filters(filters) do
-    Map.new(filters, fn {key, value} ->
+    filters
+    |> Enum.flat_map(fn {key, value} ->
       case {key, value} do
         {"id", ids} ->
-          {:ids, Params.split_on_comma(ids)}
+          %{ids: Params.split_on_comma(ids)}
 
         {"route", route_ids} ->
-          {:route_ids, Params.split_on_comma(route_ids)}
+          %{route_ids: Params.split_on_comma(route_ids)}
 
         {"direction_id", direction_id} ->
-          {:direction_id, Params.direction_id(%{"direction_id" => direction_id})}
+          %{direction_id: Params.direction_id(%{"direction_id" => direction_id})}
 
         {"stop", stop_ids} ->
-          {:stop_ids, Params.split_on_comma(stop_ids)}
+          %{stop_ids: Params.split_on_comma(stop_ids)}
 
         {"canonical", canonical} ->
-          {:canonical, Params.canonical(canonical)}
+          %{canonical: Params.canonical(canonical)}
+
+        {"date", date} ->
+          process_date(date)
       end
     end)
+    |> Enum.into(%{})
+    |> consolidate_route_ids()
+    |> gather_trip_based_route_pattern_ids()
+  end
+
+  defp consolidate_route_ids(
+         %{
+           route_ids: [_ | _] = route_ids,
+           date_specific_route_ids: [_ | _] = date_specific_route_ids
+         } = params
+       ) do
+    filtered_route_ids =
+      Enum.filter(route_ids, fn id -> Enum.member?(date_specific_route_ids, id) end)
+
+    params
+    |> Map.put(:route_ids, filtered_route_ids)
+    |> Map.delete(:date_specific_route_ids)
+  end
+
+  defp consolidate_route_ids(%{date_specific_route_ids: date_specific_route_ids} = params) do
+    params
+    |> Map.put(:route_ids, date_specific_route_ids)
+    |> Map.delete(:date_specific_route_ids)
+  end
+
+  defp consolidate_route_ids(params), do: params
+
+  defp gather_trip_based_route_pattern_ids(
+         %{date: _date, route_ids: [_ | _] = route_ids} = params
+       ) do
+    route_pattern_ids =
+      params
+      |> Map.take([:date, :direction_id])
+      |> Map.put(:routes, route_ids)
+      |> Trip.filter_by()
+      |> Enum.map(& &1.route_pattern_id)
+      |> consolidate_route_pattern_ids(params)
+
+    Map.put(params, :ids, route_pattern_ids)
+  end
+
+  defp gather_trip_based_route_pattern_ids(params), do: params
+
+  defp consolidate_route_pattern_ids(route_pattern_ids, %{ids: [_ | _] = ids}) do
+    Enum.filter(route_pattern_ids, fn rid -> Enum.member?(ids, rid) end)
```

Comment:
question: if you're doing `Enum.member?/2` checks, should `route_pattern_ids` be a `MapSet` instead of a list?